Using index mappers instead of own created storage for states of sorting

wszymanski opened the pull request:
### Context
The `ColumnSorting` plugin has `ColumnStatesManager` for managing column states. So far, it has been storing the states inside a simple native array. It was a quite fast and useful solution. However, the manager hasn't handled the situation when we reassign indexes. When we use the `alter` method with `insert_col` parameter a newly created column's index replaces the previous physical index. A similar desynchronization occurs when we removing some columns. A physical index **isn't** really a unique value.

Index maps released within `Handsontable 8.0.0` react on such actions properly, that's why we decided to change the simple array used for storing states to the instance of `IndexMap` (precisely, newly introduced `QueuedPhysicalIndexToValueMap`).

I've introduced a new kind of `IndexMap` named `QueuedPhysicalIndexToValueMap`. It maps from physical indexes to values (for the `ColumnSorting` values beings objects are assigned for particular column). This map realizes the idea of value queuing. For `ColumnSorting` (but also other plugins, e.g. `Filters`) order of states is important. This map updates the order of states right after CRUD actions on indexes.

Please note, that moving states from a native array to index mapper showed that `updateSettings` with set `columns` property hasn't been handled properly by the plugin. Sice `Handsontable 8.0.0` it should reset states (described in the [breaking change section](https://github.com/handsontable/handsontable/releases/tag/8.0.0)). Tests were changed within https://github.com/handsontable/handsontable/pull/7276/commits/5f370535d9a90dcc9e5521cc502c45956b070b20.

### How has this been tested?
It has been checked just on a config from a linked bug. Computational complexity seems to be similar for most of the actions so it should perform similarly. I've checked it on Google Chrome for more complex configs.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6397

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

budnix commented on src/translations/maps/queuedPhysicalIndexToValueMap.js:
```diff
@@ -0,0 +1,103 @@
+ * values is updated after such changes.
+ */
+class QueuedPhysicalIndexToValueMap extends IndexMap {
+  constructor() {
+    super();
+    this.queueOfIndexes = [];
+  }
```
```suggestion
queueOfIndexes = [];
```

It could be simplified by moving the prop declaration to the class fields. Then no constructor recreation is necessary. And I'd add documentation to that field.

wszymanski replied:
Done within: https://github.com/handsontable/handsontable/commit/3ac658d899570aa90829e5f048d9ce038808ed83.

budnix commented on src/translations/maps/queuedPhysicalIndexToValueMap.js:
```diff
@@ -0,0 +1,103 @@
+   */
+  remove(removedIndexes) {
+    this.indexedValues = getListWithRemovedItems(this.indexedValues, removedIndexes);
+    this.queueOfIndexes = getDecreasedIndexes(this.queueOfIndexes, removedIndexes);
```
In a case when you have for example sorted the second, the fifth, and six columns. The initial indexes are correctly stored as `[1, 4, 5]`. However, when we removing the columns where the sorting was applied (`hot.alter('remove_col', 0, 10)`) the `this.queueOfIndexes` looks like `[0, 0, 0]`. This causes weird results from the API for example `hot.getPlugin('multiColumnSorting').getSortConfig()` returns the three objects with columns `0` and adding a new column while pressing the <kbd>Ctrl/Cmd</kbd> key doesn't work anymore.

wszymanski replied:
Thanks. It should be fixed within https://github.com/handsontable/handsontable/pull/7276/commits/3ac658d899570aa90829e5f048d9ce038808ed83.

budnix commented on src/translations/maps/queuedPhysicalIndexToValueMap.js:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  /**
+   * Get every queued value.
```
```suggestion
   * Get every value in the same order as they were added to the index mapper. 
```

wszymanski replied:
I will change it if [the decision](https://github.com/handsontable/handsontable/pull/7276#discussion_r494836980) will be taken.

```suggestion
   * Get every non-default value in the same order as they were added to the index mapper. 
```

budnix commented on src/translations/maps/queuedPhysicalIndexToValueMap.js:
```diff
@@ -0,0 +1,103 @@
+ * Does not update stored values on remove/add row or column action. Otherwise, queue of indexes related to ordered
+ * values is updated after such changes.
+ */
+class QueuedPhysicalIndexToValueMap extends IndexMap {
```
I'm not convinced about the match of the word "Queued". The Queue in the definition is a structure that accepts values on the one side and removes the values on the other side - nothing happens here. Moreover IMO "queue" can suggest that something is collected for further processing (maybe asynchronous). 

I think that the better word could be "Ordered". It doesn't tell anything about the process method and still describes the meaning that the values will be accessible in the same order as they were put into. 

@wszymanski @wojciechczerniak what do you guys think about it?

wszymanski replied:
Leaving aside the fact that queue is also a term used for describing things beyond programming (and for me, it may be still useful for naming abstract beings; I use sometimes, for example, the phrase `list` for naming things not necessarily implementing the list in the sense of programming) I would like to pay attention to the fact there are different types of queue 🙂 

In programming, we also distinguish for example **priority queue**. We add elements to an array using priority starting from elements with higher priority and ending with elements with lower priority. 

Anyway, I remember that in a discussion with @wojciechczerniak there were also remarks about the name. If you both agree about using another phrase for the `IndexMap` and elements within it I will change it.

@wojciechczerniak What is your opinion?

wojciechczerniak replied:
I agree that the name does not feel right. But I don't have a better one. 

It's hard to go beyond programming, we're programming here 😄 
Queue has a definition: https://www.hackerearth.com/practice/data-structures/queues/basics-of-queues/tutorial/

"Queued", "Ordered", "Sparse", "Dense", "Indexed", "Weighted" 🤔 

I think that the closest to what we have would be a Linked Hash Map? Physical indexes to value are a hash map, but we have an additional information about their order -> links. 
I found this that looks exactly what we did: 
![hash_maps](https://user-images.githubusercontent.com/4718998/95028786-24193600-06a3-11eb-95cc-3bfa9f81cdae.png)
But maybe I'm just seeing things that proof my idea. WDYT?

Maybe `LinkedPhysicalIndexToValueMap` ?

wszymanski replied:
> I agree that the name does not feel right. But I don't have a better one.

We create something more abstract than just the regular data structure. Just some of the values are ordered/queued/whatever. I think whether any of names represent that fact 🤔 

> It's hard to go beyond programming, we're programming here 😄
> Queue has a definition: [hackerearth.com/practice/data-structures/queues/basics-of-queues/tutorial](https://www.hackerearth.com/practice/data-structures/queues/basics-of-queues/tutorial/)

I've mentioned just that as the `queue` is quite a good word for ordered elements from `0` to `n`. Literally, it sounds good for me. But I haven't focused too much on it, let's put it aside. Following the example of the **priority queue**, please take a look at the definition:

In computer science, a priority queue is an abstract data type similar to regular queue or stack data structure in which each element additionally has a "priority" associated with it. **In a priority queue, an element with high priority is served before an element with low priority**. In some implementations, if two elements have the same priority, they are served according to the order in which they were enqueued, while in other implementations, ordering of elements with the same priority is undefined.

While priority queues are often implemented with heaps, they are conceptually distinct from heaps. **A priority queue is a concept like "a list" or "a map"**; just as a list can be implemented with a linked list or an array, a priority queue can be implemented with a heap or a variety of other methods such as an unordered array.

> I think that the closest to what we have would be a Linked Hash Map? Physical indexes to value are a hash map, but we have an additional information about their order -> links.
> I found this that looks exactly what we did:
> ![hash_maps](https://user-images.githubusercontent.com/4718998/95028786-24193600-06a3-11eb-95cc-3bfa9f81cdae.png)
> But maybe I'm just seeing things that proof my idea. WDYT?
> 
> Maybe `LinkedPhysicalIndexToValueMap` ?

It sounds reasonable for me, WDYT @budnix?

budnix replied:
> It sounds reasonable for me, WDYT @budnix?

I would choose the `OrderedPhysicalIndexToValueMap` name 😉  The `LinkedPhysicalIndexToValueMap ` name at first glance doesn't say anything about the business logic of that specific IndexMap class. However, after looking into the source code I am able to find a connotation with the `linked` name. Summarizing if you don't feel that the `Ordered...` class name doesn't describe the module behavior enough I'm open to @wojciechczerniak suggestion and we can use `Linked...` 👍

budnix commented on handsontable.d.ts:
```diff
@@ -1974,7 +1974,7 @@ declare namespace Handsontable {
   namespace RecordTranslation {
     interface IndexMap {
-      getValues: number[],
+      getValues: () => any[],
```
```suggestion
      getValues: () => any[] | any[],
```

budnix replied:
Does it seem that the method returns not only functions with different return signatures but returns different primitives as well as?

budnix commented:
Referring to the comment https://github.com/handsontable/handsontable/pull/7276#discussion_r494825123.

The first demo shows that after removing the two first columns the rest of them wrongly indicates the number in the header
![Kapture 2020-09-25 at 10 49 29](https://user-images.githubusercontent.com/571316/94247218-1cc49080-ff1d-11ea-9545-f9c23521a63f.gif)

And the second gif shows that after removing the first columns we cannot sort other columns. Moreover, API returns doubled sort state.
![Kapture 2020-09-25 at 10 50 36](https://user-images.githubusercontent.com/571316/94247556-9492bb00-ff1d-11ea-8780-bb9b40585a75.gif)

I hope that this will help you in issue investigation 🕵️

wojciechczerniak commented on src/plugins/columnSorting/columnStatesManager.js:
```diff
@@ -156,50 +135,72 @@ export class ColumnStatesManager {
    */
   getColumnSortState(column) {
     if (this.isColumnSorted(column)) {
-      return deepClone(this.sortedColumnsStates[this.getIndexOfColumnInSortQueue(column)]);
+      const physicalColumn = this.hot.toPhysicalColumn(column);
+
+      const sortingStateWithPriority = this.sortingStates.getValueAtIndex(physicalColumn);
```
If condition `isColumnSorted` internally does `isObject(this.sortingStates.getValueAtIndex(this.hot.toPhysicalColumn(column)))` then again we call `toPhysicalColumn`, and another `getValueAtIndex`. Those are cheap but unnecessary.

We can move the check to the end
```
if (isObject(sortingStateWithPriority)) {
  return {
    ...
  }
}
```

wszymanski replied:
Changed within https://github.com/handsontable/handsontable/pull/7276/commits/6ec835fef3ffef5d061ac43e1aabc5f5b834c2bf with some other improvements.

wojciechczerniak reviewed (changes requested):
I've noticed few things. 

Why do we create an array with single element just to slice it and take a single element? `[sortConfig].slice(0, 1)`. A copy is needed?
https://github.com/handsontable/handsontable/blob/3ac658d899570aa90829e5f048d9ce038808ed83/src/plugins/columnSorting/columnSorting.js#L319-L325

We can exit early here. We already have `onlyExistingVisualIndexes` in hand. If it's `false` we can exit, no need to run `areValidSortStates(sortConfigs)` then.
https://github.com/handsontable/handsontable/blob/3ac658d899570aa90829e5f048d9ce038808ed83/src/plugins/columnSorting/columnSorting.js#L339-L344
And `const sortedColumns = sortConfigs.map(({ column }) => column);` is rather unnecessary loop. We can check `every` without flattening the object.

Inside `areValidSortStates` (utils.js) we flatten the `sortConfigs.map(({ column }) => column)` again. Moreover, we iterate over `sortConfigs` three times, one `map`, two `every`. From the first call to `areValidSortConfigs` we're iterating over the `sortConfigs` five times already?

wszymanski commented:
> I've noticed few things.
> 
> Why do we create an array with single element just to slice it and take a single element?

It's not part of this PR. However, I can't see any purpose in this part of the code. I removed it within https://github.com/handsontable/handsontable/pull/7276/commits/0cccee42301d9cbf5226c12fd6a759db911d256b.

https://github.com/handsontable/handsontable/blob/3ac658d899570aa90829e5f048d9ce038808ed83/src/plugins/columnSorting/columnSorting.js#L319-L325
> 
> We can exit early here. We already have `onlyExistingVisualIndexes` in hand. If it's `false` we can exit, no need to run `areValidSortStates(sortConfigs)` then.

I've refactored it within https://github.com/handsontable/handsontable/pull/7276/commits/97a3fb83128efaf49c0012608d30a6f3cd92d82b.

> https://github.com/handsontable/handsontable/blob/3ac658d899570aa90829e5f048d9ce038808ed83/src/plugins/columnSorting/columnSorting.js#L339-L344
> 
> 
> And `const sortedColumns = sortConfigs.map(({ column }) => column);` is rather unnecessary loop. We can check `every` without flattening the object.

Sure.

> Inside `areValidSortStates` (utils.js) we flatten the `sortConfigs.map(({ column }) => column)` again. 

But, that part of the code seems to be needed. It is a fast check of the uniqueness of indexes.

> Moreover, we iterate over `sortConfigs` three times, one `map`, two `every`. From the first call to `areValidSortConfigs` we're iterating over the `sortConfigs` five times already?

It has been reduced to three iterations. Removing `utils` would also decrease the number of loops by one. However, I've left it untouched.

wszymanski commented on src/plugins/columnSorting/columnSorting.js:
```diff
@@ -226,17 +229,13 @@ class ColumnSorting extends BasePlugin {
     }
 
-    this.hot.runHooks('afterColumnSort', currentSortConfig, this.getSortConfig(), sortPossible);
+    this.hot.runHooks('afterColumnSort',
+      currentSortConfig, sortPossible ? destinationSortConfigs : currentSortConfig, sortPossible);
```
So two calls for `runHooks` - one within `if` and second one within `else`? I am not sure I understand your idea.

wojciechczerniak replied:
No. `sortPossible ? destinationSortConfigs : currentSortConfig` is an if that we have two lines higher with ` if (sortPossible) {` so my proposition is to add `currentSortConfig = destinationSortConfigs;` inside and call `runHooks` only once.

```suggestion
   if (sortPossible) {
      this.columnStatesManager.setSortStates(destinationSortConfigs);
      this.sortByPresetSortStates(destinationSortConfigs);
      this.saveAllSortSettings(destinationSortConfigs);
      
      // here --------------
      currentSortConfig = destinationSortConfigs;
    }

    this.hot.runHooks('afterColumnSort', currentSortConfig, currentSortConfig, sortPossible);
```

wojciechczerniak replied:
BTW. it's kind of weird that we call it with two `currentSortConfigs`. Wouldn't `destinationConfig === null` make more sense if `sortPossible === false`? There is no destination config then 🤔 But that would be a breaking change so don't mind me. Just thinking loud 🙄

wszymanski replied:
The `afterColumnSort` hook is called with start config and destination config. Your idea would change that. We would lose information about start config by the assignee.

I don't feel that `destinationConfig` should be `null`. Information on what config user tried to reach may be useful.

wojciechczerniak replied:
You're right. That will overwrite both values and introduce a bug. Sorry

wojciechczerniak commented:
> It's not part of this PR.

Yeah, I know. Sorry about that. I'm trying to look at the big picture here. After changing the structure some operations may be redundant or harmful to the performance. Thanks for taking those under consideration and all the fixes.

wszymanski commented on src/plugins/columnSorting/columnStatesManager.js:
```diff
@@ -84,55 +93,25 @@ export class ColumnStatesManager {
   getIndexOfColumnInSortQueue(column) {
-    return this.getSortedColumns().indexOf(column);
+    return this.sortingStates.getIndexesQueue()
+      .findIndex(physicalColumn => this.hot.toVisualColumn(physicalColumn) === column);
```
Changed within https://github.com/handsontable/handsontable/pull/7276/commits/f9bfe4a4837707b1a72e37d058bc1306cdff54a1.

wszymanski commented on src/translations/maps/queuedPhysicalIndexToValueMap.js:
```diff
@@ -0,0 +1,107 @@
+   *
+   * @returns {Array}
+   */
+  getIndexesQueue() {
+    return this.queueOfIndexes;
+  }
```
I understand the idea, but I'm afraid that we shouldn't focus mainly on the possibility of redefining the map. The map provides new functionality as part of a **new structure**.

If it introduces new features thus it may have a new API for newly introduced parts of the functionality. This feature introduces information that **just some indexes** have non-default values and those **entries** are arranged in the order. That's why we introduced `getEntries` without overriding the old `getValues` method. Consistently I can't see a problem with introducing `getIndexesQueue` method (somehow it will be called finally).

Some methods can't be overridden as features of the "base" Index Map still may be used and I'm afraid that it should be still provided. The `getLength` method still represents the full size of the Index Map (also with "empty" values).

I'm concerned whether `clear` method should be overridden 🤔 .

wojciechczerniak replied:
If we're going to use it as a LinkedMap, then the length (and therefore `getLength` value) is the number of linked elements. We have no interest in the empty values in this case. IMO `getValues` may be overwritten as well. Why would we need it in the current form if were interested in `queueOfIndexes` only?

wszymanski replied:
Hmm... I wrote [here](https://github.com/handsontable/handsontable/pull/7276#discussion_r499690664) that we create something unusual. From my perspective it look like:

![Untitled Diagram (11)](https://user-images.githubusercontent.com/141330/95203505-24204f80-07e3-11eb-9fd4-05574e72f9bd.jpg)

It's kind of hybrid where some values are just in the hash map, and part of them (it is important that it does not apply to all of them) are stored inside linked list.

Do we consider override of `getValueAtIndex` method? Now, we don't. Why? Because we are not dealing only a linked list (or precisely, Linked HashSet).

wojciechczerniak replied:
Not that unusual ;) Look: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

The Map object has `Map.prototype.get(key)` that will return the value from the "Hash Set", so if the key does not exist it will return `undefined` just as our `getValueAtIndex` will return `null`.

At the same time Map object has `Map.prototype.values()` method that will return ONLY the values from the "Linked List". Moreover, it says "values for each element in the Map object **in insertion order**." so basically Javascript Map object is the Linked Hash Map.

IMO looks just like what you draw.

I agree with @budnix that we miss something from "Linked" structure. It's the iterator, you can't get `next()` linked item, just all `entries()` in the correct order. Something we can add later. Ideally, this IndexMap could be internally based on `Map.prototype`, not an `Array.prototype` so we have all of those out of the box and we would not need additional `queueOfIndexes` to store the order, JS does that for us.

wszymanski replied:
Okay, it convinces me. I can see that I incorrectly mixed current implementation with independent modeling of being. I will do my best to refactor it. Thanks 👍

budnix commented on test/unit/translations/maps/linkedPhysicalIndexToValueMap.spec.js:
```diff
@@ -0,0 +1,217 @@
+  indexToValueMap.init(3);
+
+  expect(indexToValueMap.indexedValues).toEqual([{ key: 0 }, { key: 1 }, { key: 2 }]);
+  expect(indexToValueMap.getValues()).toEqual([]); // Returns only non-default values.
```
The comment `// Returns only non-default values.` can be confusing here. As in the line above the `indexedValues` are filled with some non-null values which can be read as non-default ones. But in fact, `getValues` returns an empty array because for that Collection the `orderOfIndexes` array is a key.

wszymanski replied:
Removed within: https://github.com/handsontable/handsontable/pull/7276/commits/069622ac53b024960a5510a69df4b19e42cb672f and https://github.com/handsontable/handsontable/pull/7276/commits/8494f6bf9c2adba659f81b6bf8a60cc43a95b8d8

budnix commented on test/unit/translations/maps/linkedPhysicalIndexToValueMap.spec.js:
```diff
@@ -0,0 +1,217 @@
+import { LinkedPhysicalIndexToValueMap as IndexToValueMap } from 'handsontable/translations';
```
I think that the unit test for that kind of IndexMap could be improved. I've found that some methods aren't tested very well or not tested at all. For example, I removed `getLength`, `insert` and `remove` methods from the LinkedMap, and run test specs defined by that particular file - all test passes (the error was caught within IndexMapper specs). When I removed only the `getLength` method all unit test passes.

Generally speaking, I'd add a unit test for each method separately by checking the internal state after execution. This collection is especially important as the new internal state `orderOfIndexes` is changing the Map abstract behavior.

wszymanski replied:
Thanks for the critical remarks 👍 The generic `getLength` will work, but it's slower. That's the reason why it's overridden. Thus, removing it won't give effects in unit tests. However, I've added an extra check for it.

budnix reviewed (changes requested):
Good Job 👍  I've left my thoughts in the code comments.

wojciechczerniak reviewed (dismissed):
Nothing to add, +1 on @budnix remarks. 👍 from me. Awesome 🥇

budnix reviewed (approved):
🏅

aninde commented:
Two tests are not passing on my latest Chrome 86 / macOS Catalina after these changes. 
<img width="925" alt="Screenshot 2020-10-26 at 15 24 20" src="https://user-images.githubusercontent.com/30186818/97184964-c9559480-179f-11eb-8160-54a9bec8f02c.png">

wszymanski commented:
Could you build distribution files once again and clear a browser cache? I've used the same config and it seems to work.

aninde commented:
@wszymanski you're right, I confirm that all tests from `columnSorting/test/columnSorting.e2e.js` pass after running them on the clean environment on branch `develop`.

<img width="193" alt="Screenshot 2020-10-27 at 15 37 49" src="https://user-images.githubusercontent.com/30186818/97316769-6977f000-186a-11eb-85f0-ac6a32afe3c7.png">
